Automattic/jetpack
# CRM: Ensuring the assigned owner of a contact is not removed upon editing the contact

Fixes https://github.com/Automattic/zero-bs-crm/issues/3438

## Proposed changes:

* ~This PR makes sure a value is added to the assigned owner meta box when a user has been assigned, so that upon saving after editing the contact that value is not replaced.~
* This PR ensures only contact owners who have permission to change owners can actually change the owner, by checking their capabilities within the save function itself, and then ensuring that if they are already the owner they remain the current owner.
* This prevents the issue of assigned users being unassigned from contacts after editing the contact (if 'Assign Ownership' is unchecked in the CRM general settings page), which was happening because a value for the user id wasn't being saved upon contact edit.

### Other information:


## Jetpack product discussion

https://github.com/Automattic/zero-bs-crm/issues/3438

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

To replicate the issue: 

- Create a couple of test contacts
- Create a CRM Contact Manager user
- Assign a contact to that user
- Log in as that user, and attempting to make any change to that contact (eg. status).
- The change should stick, but the user is now unassigned from that contact

To test the fix, with this PR applied:

- With the same test contacts and user assignments as above, ensure you can edit a contact for which you have been assigned, and that your assignment isn't removed when doing so
- Make sure you can't edit a contact for which you haven't been assigned (unless it is unassigned).
- Also, make sure you cannot select any other user to be the owner of that contact.
- To make sure there are no regressions:
- Change the settings to make sure 'Assign Ownership' is checked, and ensure you can edit any contact, regardless of whether you are assigned or someone else is. Make sure you can also change the owner.
- You can also do the above testing for any company that has been assigned to a user as well.
